frontend/lib/src/components/widgets/TextInput/TextInput.tsx

```diff
@@ -144,7 +144,20 @@ class TextInput extends React.PureComponent<Props, State> {
     // If the TextInput is *not* part of a form, we mark it dirty but don't
     // update its value in the WidgetMgr. This means that individual keypresses
     // won't trigger a script re-run.
-    this.setState({ dirty: true, value })
+    if (!isInForm(this.props.element)) {
+      this.setState({ dirty: true, value })
+      return
+    }
+
+    // If TextArea *is* part of a form, we immediately update its widgetValue
+    // on text changes. The widgetValue won't be passed to the Python
+    // script until the form is submitted, so this won't cause the script
+    // to re-run.
+    this.setState({ dirty: true, value }, () =>
```

Comment:
Setting `dirty` to true here seems off to me since you then pass in arg to `commitWidgetValue` to set it to `true` a second time within that function. I'd guess this is a sign that some more refactoring is probably needed to avoid this kind of clumsiness

Reply:
Good point. I'll refactor this because setting dirty but committing definitely seems odd.